### PR TITLE
feat: added AppCustom configuration to ds-controller-board

### PR DESCRIPTION
**Pull request by marcpfuller**
Added AppCustom configuration and moved Driver config to AppCustom section in configuration toml. This resolves issue #25 for ds-controller-board

Signed-off-by: Marc-Philippe Fuller <marc-philippe.fuller@intel.com>

### CODE REVIEW

File: ds-controller-board/res/configuration.toml [hunk 1]
```diff
@@ -85,4 +78,12 @@ PublishTopicPrefix = "edgex/events/device" # /<device-profile-name>/<device-name
   Retained = "false"
   AutoReconnect = "true"
   ConnectTimeout = "5" # Seconds
-  SkipCertVerify = "false"
\ No newline at end of file
+  SkipCertVerify = "false"
+
+[AppCustom]
+  [DriverConfig]
```
Comment (lenny-intel) [inline ds-controller-board/res/configuration.toml:84]: Why need for extra level. I would just have the settings under `[AppCustom]` or rename `[AppCustom]` to `[DriverConfig]`. The name AppCustom originated for an example, but can be anything...

File: ds-controller-board/res/configuration.toml [hunk 2]
```diff
@@ -85,4 +78,12 @@ PublishTopicPrefix = "edgex/events/device" # /<device-profile-name>/<device-name
...
+  VID = '2341'
+  PID = '8037'
+  DisplayTimeout = '10s'
+  LockTimeout = '30s'
+  VirtualControllerBoard = 'true'
```
Comment (lenny-intel) [inline ds-controller-board/res/configuration.toml:89]: The reason to make this custom config is we can have typed data rather than all strings.
```suggestion
  VirtualControllerBoard = true
```

File: ds-controller-board/res/configuration.toml [hunk 3]
```diff
@@ -85,4 +78,12 @@ PublishTopicPrefix = "edgex/events/device" # /<device-profile-name>/<device-name
...
   AutoReconnect = "true"
   ConnectTimeout = "5" # Seconds
-  SkipCertVerify = "false"
\ No newline at end of file
+  SkipCertVerify = "false"
+
+[AppCustom]
+  [DriverConfig]
+  VID = '2341'
+  PID = '8037'
```
Comment (lenny-intel) [inline ds-controller-board/res/configuration.toml:86]: Use `"` instead o `'` here and any other places. `"` are the standard. `'` are use in special cases when the string is wrapped across lines.

Comment (lenny-intel) [inline ds-controller-board/res/configuration.toml:86]: I think VID and PID should stay as strings correct. But, they really should be moved to the protocol section of the Device definition. But that can be a separate PR.

File: ds-controller-board/device/config.go [hunk 4]
```diff
@@ -4,15 +4,38 @@
...
 )
 
+// CustomConfig holds the values for the driver configuration
+type CustomConfig struct {
+	DriverConfig Config
+}
+
+// ServiceConfig a struct that wraps CustomConfig which holds the values for driver configuration
+type ServiceConfig struct {
+	AppCustom CustomConfig
```
Comment (lenny-intel) [inline ds-controller-board/device/config.go:17]:
```suggestion
	DriverConfig Config
```

File: ds-controller-board/device/config.go [hunk 5]
```diff
@@ -4,15 +4,38 @@
...
 
 import (
-    "time"
+	"time"
 )
 
+// CustomConfig holds the values for the driver configuration
+type CustomConfig struct {
+	DriverConfig Config
+}
```
Comment (lenny-intel) [inline ds-controller-board/device/config.go:13]: Remove unneeded level

File: ds-controller-board/device/config.go [hunk 6]
```diff
@@ -4,15 +4,38 @@
...
+
 // Config is the global device configuration, which is populated by values in
 // the "Driver" section of res/configuration.toml
 type Config struct {
-    VirtualControllerBoard bool
-    PID                    string
-    VID                    string
-    DisplayTimeout         time.Duration
-    LockTimeout            time.Duration
+	VirtualControllerBoard bool
```
Comment (lenny-intel) [inline ds-controller-board/device/config.go:23]: Perfect

File: ds-controller-board/device/config.go [hunk 7]
```diff
@@ -4,15 +4,38 @@
...
+	PID                    string
+	VID                    string
+	DisplayTimeout         time.Duration
+	LockTimeout            time.Duration
```
Comment (lenny-intel) [inline ds-controller-board/device/config.go:27]: So will the TOML marshaler properly marshal the strings into these  time.Duration fields?
Please try putting invalid duration values to see how it handles it.

File: ds-controller-board/device/device.go [hunk 8]
```diff
@@ -20,13 +20,13 @@ type ControllerBoard interface {
 // NewControllerBoard is used to determine the ControllerBoard type
 // (i.e. Device or Virtual), and perform the necessary steps to initialize
 // a new ControllerBoard
-func NewControllerBoard(lc logger.LoggingClient, asyncCh chan<- *dsModels.AsyncValues, config *Config) (ControllerBoard, error) {
+func NewControllerBoard(lc logger.LoggingClient, asyncCh chan<- *dsModels.AsyncValues, config *ServiceConfig) (ControllerBoard, error) {
 	var controllerBoard ControllerBoard
 
-	if !config.VirtualControllerBoard {
+	if !config.AppCustom.DriverConfig.VirtualControllerBoard {
```
Comment (lenny-intel) [inline ds-controller-board/device/device.go:26]:
```suggestion
	if !config.VirtualControllerBoard {
```

File: ds-controller-board/device/device.go [hunk 9]
```diff
@@ -20,13 +20,13 @@ type ControllerBoard interface {
 // NewControllerBoard is used to determine the ControllerBoard type
 // (i.e. Device or Virtual), and perform the necessary steps to initialize
 // a new ControllerBoard
-func NewControllerBoard(lc logger.LoggingClient, asyncCh chan<- *dsModels.AsyncValues, config *Config) (ControllerBoard, error) {
+func NewControllerBoard(lc logger.LoggingClient, asyncCh chan<- *dsModels.AsyncValues, config *ServiceConfig) (ControllerBoard, error) {
```
Comment (lenny-intel) [inline ds-controller-board/device/device.go:23]: Could pass in config.DriverConfig to shorten the lines below
```suggestion
func NewControllerBoard(lc logger.LoggingClient, asyncCh chan<- *dsModels.AsyncValues, config *DriverConfig) (ControllerBoard, error) {
```

File: ds-controller-board/device/device.go [hunk 10]
```diff
@@ -20,13 +20,13 @@ type ControllerBoard interface {
...
 	var controllerBoard ControllerBoard
 
-	if !config.VirtualControllerBoard {
+	if !config.AppCustom.DriverConfig.VirtualControllerBoard {
 
 		// Find the port name (like /dev/ttyACM0) connected to Controller Board
-		ttyPort, err := FindControllerBoard(config.VID, config.PID)
+		ttyPort, err := FindControllerBoard(config.AppCustom.DriverConfig.VID, config.AppCustom.DriverConfig.PID)
```
Comment (lenny-intel) [inline ds-controller-board/device/device.go:29]:
```suggestion
		ttyPort, err := FindControllerBoard(config.VID, config.PID)
```

File: ds-controller-board/driver/driver.go [hunk 11]
```diff
@@ -52,10 +56,25 @@ func (drv *ControllerBoardDriver) Initialize(lc logger.LoggingClient, asyncCh ch
...
-		if err = utilities.MarshalSettings(service.DriverConfigs(), drv.config, true); err != nil {
-			return err
+		drv.svc = &DeviceSDKService{
+			DeviceService: service.RunningService(),
+			lc:            lc,
+		}
+
+		drv.config = &device.ServiceConfig{}
+
+		err := drv.svc.LoadCustomConfig(drv.config, "AppCustom")
```
Comment (lenny-intel) [inline ds-controller-board/driver/driver.go:66]:
```suggestion
		err := drv.svc.LoadCustomConfig(drv.config, "DriverConfig")
```

File: ds-controller-board/driver/driver.go [hunk 12]
```diff
@@ -52,10 +56,25 @@ func (drv *ControllerBoardDriver) Initialize(lc logger.LoggingClient, asyncCh ch
...
+		if err != nil {
+			return errors.Wrap(err, "custom driver configuration failed to load")
+		}
+
+		drv.lc.Debugf("Custom config is : %+v", drv.config)
```
Comment (lenny-intel) [inline ds-controller-board/driver/driver.go:71]:
```suggestion
		drv.lc.Debugf("Custom driver config is : %+v", drv.config)
```

File: ds-controller-board/driver/driver.go [hunk 13]
```diff
@@ -52,10 +56,25 @@ func (drv *ControllerBoardDriver) Initialize(lc logger.LoggingClient, asyncCh ch
...
+
+		err = drv.svc.ListenForCustomConfigChanges(&drv.config.AppCustom, "AppCustom", drv.updateWritableConfig)
+		if err != nil {
+			return errors.Wrap(err, "failed to listen to custom config changes")
 		}
+
```
Comment (lenny-intel) [inline ds-controller-board/driver/driver.go:77]: Do we need this config to be "writable"? I don't think so...

File: ds-controller-board/driver/driver.go [hunk 14]
```diff
@@ -107,7 +126,7 @@ func (drv *ControllerBoardDriver) HandleWriteCommands(deviceName string, protoco
 				return err
 			}
 			go func() {
-				time.Sleep(drv.config.LockTimeout)
+				time.Sleep(drv.config.AppCustom.DriverConfig.LockTimeout)
```
Comment (lenny-intel) [inline ds-controller-board/driver/driver.go:129]:
```suggestion
				time.Sleep(drv.config.DriverConfig.LockTimeout)
```

File: ds-controller-board/driver/driver.go [hunk 15]
```diff
@@ -126,7 +145,7 @@ func (drv *ControllerBoardDriver) HandleWriteCommands(deviceName string, protoco
...
```
Comment (lenny-intel) [inline ds-controller-board/driver/driver.go:148]:
```suggestion
				time.Sleep(drv.config.DriverConfig.LockTimeout)
```

File: ds-controller-board/driver/driver.go [hunk 16]
```diff
@@ -219,7 +238,7 @@ func (drv *ControllerBoardDriver) displayText(message string) {
 	go func() {
 		for {
 			select {
-			case <-time.After(drv.config.DisplayTimeout):
+			case <-time.After(drv.config.AppCustom.DriverConfig.DisplayTimeout):
```
Comment (lenny-intel) [inline ds-controller-board/driver/driver.go:241]:
```suggestion
			case <-time.After(drv.config.DriverConfig.DisplayTimeout):
```

File: ds-controller-board/driver/driver.go [hunk 17]
```diff
@@ -260,3 +279,24 @@ func (drv *ControllerBoardDriver) RemoveDevice(deviceName string, protocols map[
 func (drv *ControllerBoardDriver) Stop(force bool) error {
 	return nil
 }
+
+func (drv *ControllerBoardDriver) updateWritableConfig(rawWritableConfig interface{}) {
+	updated, ok := rawWritableConfig.(*device.CustomConfig)
+	if !ok {
+		drv.lc.Error("unable to update writable custom config: Can not cast raw config to type 'CustomConfig'")
```
Comment (lenny-intel) [inline ds-controller-board/driver/driver.go:286]: VID and PID need to be move into the Device Protocol. So this is not needed.
If we allows writable you have to consider how the value is used and if there is special processing needed when the value changes. IMO, these settings are not going to need to change at run time to justify this complexity.

File: ds-controller-board/driver/driver_test.go [hunk 18]
```diff
@@ -156,8 +156,12 @@ func CreateControllerBoardDriver(t *testing.T, virtual bool, initialize bool, ex
...
 	target := &ControllerBoardDriver{
 		lc: lc,
-		config: &device.Config{
-			VirtualControllerBoard: virtual,
+		config: &device.ServiceConfig{
+			AppCustom: device.CustomConfig{
+				DriverConfig: device.Config{
+					VirtualControllerBoard: virtual,
+				},
+			},
```
Comment (lenny-intel) [inline ds-controller-board/driver/driver_test.go:164]:
```suggestion
		       DriverConfig: device.Config{
			       VirtualControllerBoard: virtual,
			},
```

File: ds-controller-board/driver/service.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"github.com/edgexfoundry/device-sdk-go/v2/pkg/service"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/clients/logger"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+	"github.com/pkg/errors"
+)
+
+// ServiceWrapper wraps an EdgeX SDK service so it can be easily mocked in tests.
+type ServiceWrapper interface {
```
Comment (lenny-intel) [inline ds-controller-board/driver/service.go:11]: Why add this?

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: ds-controller-board/driver/service.go (hunk 19 above)
Comment (marcpfuller) [reply ds-controller-board/driver/service.go:11]: the intefaces module is not in the device SDK 2.2.0 version

Comment (marcpfuller) [reply ds-controller-board/driver/service.go:11]: I needed the load custom config. thats why I added the Service wrapper following how device LLRP was before I removed it in Edgex for 2.3

Comment (lenny-intel) [reply ds-controller-board/driver/service.go:11]: load custom config is available without this. Only reason to add this is so you can mock the API in unit tests, which we are not adding. LoadCustomConfig is available from object return by `service.RunningService()`

File: ds-controller-board/driver/service.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+	return service.DriverConfigs()
+}
+
+func (s *DeviceSDKService) SetDeviceOpState(name string, state contract.OperatingState) error {
+	// workaround: the device-service-sdk's uses core-contracts for the API URLs,
+	// but the metadata service API for opstate updates changed between v1.1.0 and v1.2.0.
+	d, err := s.GetDeviceByName(name)
+	if err != nil {
+		return errors.New("no such device")
+	}
```
Comment (jim-wang-intel) [inline ds-controller-board/driver/service.go:46]: should include the err in the log?
```suggestion
	if err != nil {
		return fmt.Errorf("found error to get device by name: %v", err)
	}
```

File: ds-controller-board/driver/driver.go (hunk 13 above)
Comment (brian-intel) [reply ds-controller-board/driver/driver.go]: The 2 timeouts could be modified in real time but it's unlikely
Simulated, pid, and vid require a service reboot to cleanly connect to the device.

File: ds-controller-board/driver/driver.go (hunk 17 above)
Comment (brian-intel) [reply ds-controller-board/driver/driver.go:283]: Correct, VID, PID, and simulated will require a service reboot.

File: ds-controller-board/driver/service.go (hunk 20 above)
Comment (marcpfuller) [reply ds-controller-board/driver/service.go:46]: Based on the changes Lenny suggested, service.go is not needed. So I have removed it.

**Review by brian-intel — APPROVED** [review]

File: ds-controller-board/device/config_test.go [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+			ok := actualConfig.UpdateFromRaw(testCase.rawConfig)
+
+			assert.Equal(t, testCase.isValid, ok)
+			if testCase.isValid {
+				assert.Equal(t, expectedConfig, &actualConfig)
+			}
+		})
+	}
+
+}
```
Comment (lenny-intel) [inline ds-controller-board/device/config_test.go:53]: Need test for validate also

File: ds-controller-board/device/config.go [hunk 22]
```diff
@@ -4,15 +4,49 @@
...
+func (c *ServiceConfig) Validate() (time.Duration, time.Duration, error) {
+
+	displayTimeout, err := time.ParseDuration(c.DriverConfig.DisplayTimeout)
+	if err != nil {
+		return 0, 0, fmt.Errorf("display timeout configuration is not a proper time duration")
+	}
+
+	lockTimeout, err := time.ParseDuration(c.DriverConfig.LockTimeout)
+	if err != nil {
+		return 0, 0, fmt.Errorf("lock timeout configuration is not a proper time duration")
```
Comment (jim-wang-intel) [inline ds-controller-board/device/config.go:48]:
```suggestion
		return 0, 0, fmt.Errorf("lock timeout configuration is not a proper time duration: %v", err)
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Review by jim-wang-intel — APPROVED** [review]

**Review by joshua-silverio — APPROVED** [review]
lgtm

File: ds-controller-board/device/config_test.go [hunk 23]
```diff
@@ -0,0 +1,110 @@
+// Copyright © 2020 Intel Corporation. All rights reserved.
```
Comment (vli11) [inline ds-controller-board/device/config_test.go:1]: update the copyright year

**Review by vli11 — CHANGES_REQUESTED** [review]
